tests/components/alarmdecoder/test_config_flow.py

```diff
@@ -0,0 +1,424 @@
+            {
+                CONF_DEVICE_PATH: "/dev/ttyUSB123",
+                CONF_DEVICE_BAUD: 115000,
+            },
+            115000,
+            "/dev/ttyUSB123",
+        ),
+    ],
+)
+async def test_setups(hass: HomeAssistant, protocol, connection, baud, title):
+    """Test flow for setting up the available AlarmDecoder protocols."""
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {CONF_PROTOCOL: protocol},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "protocol"
+
+    with patch("homeassistant.components.alarmdecoder.config_flow.AdExt.open"), patch(
+        "homeassistant.components.alarmdecoder.config_flow.AdExt.close"
+    ), patch(
+        "homeassistant.components.alarmdecoder.async_setup", return_value=True
+    ) as mock_setup, patch(
+        "homeassistant.components.alarmdecoder.async_setup_entry",
+        return_value=True,
+    ) as mock_setup_entry:
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], connection
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_CREATE_ENTRY
+        assert result["title"] == title
+        assert result["data"] == {
+            **connection,
+            CONF_PROTOCOL: protocol,
+            CONF_DEVICE_BAUD: baud,
+        }
+
+    await hass.async_block_till_done()
+    assert len(mock_setup.mock_calls) == 1
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_setup_connection_error(hass: HomeAssistant):
+    """Test flow for setup with a connection error."""
+
+    port = 1001
+    host = "alarmdecoder"
+    protocol = PROTOCOL_SOCKET
+    connection_settings = {CONF_HOST: host, CONF_PORT: port}
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {CONF_PROTOCOL: protocol},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "protocol"
+
+    with patch(
+        "homeassistant.components.alarmdecoder.config_flow.AdExt.open",
+        side_effect=NoDeviceError,
+    ), patch("homeassistant.components.alarmdecoder.config_flow.AdExt.close"):
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], connection_settings
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["errors"] == {"base": "service_unavailable"}
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6158285)